sycl/doc/extensions/proposed/sycl_ext_codeplay_kernel_fusion.asciidoc

```diff
@@ -0,0 +1,550 @@
+  execution of other command-groups, if the command-groups synchronize
+  via requirements or explicit synchronization. In this context, a
+  circular dependencies arise if any kernel in a fusion list depends
+  on a kernel submitted for execution in a different queue and, at the
+  same time, this depends on another kernel in the fusion list. This
+  causes a circular dependency as the fused kernel would depend on the
+  kernel not in the fusion list and, at the same time, this would
+  depend on the fused kernel.
++
+Circular dependencies can be caused by device kernels, host tasks or
+explicit memory operations. Implementations must cancel fusion in time
+to avoid such circular dependencies and deadlock of the
+application. The concrete event/submission causing cancellation is
+implementation defined. Implementations could opt to cancel only when
+the submission would create a circular dependency, but are free to do
+so earlier, e.g., on submission of a command-group to another queue
+which synchronizes with a kernel in the fusion list of another queue.
+* _Queue operations_: Calls to queue operations blocking execution of the
+calling thread, such as `sycl::queue::wait()`, must also imply an implicit
+kernel fusion cancellation.
+* _SYCL event objects_: Host synchronization on events returned by a call to
+`queue::submit` while the queue is still in fusion mode would also result on an
+implicit kernel fusion cancellation.  Explicit dependencies (specified by the
+user with `handler::depends_on`) between kernels to be fused must be dropped, as
+the requirement will trivially hold (per work-item) thanks to fusion semantics.
+* _Queue destruction_: As in this extension the queue becomes stateful, the
+destruction of a queue in fusion mode would lead to an implicit kernel fusion
+cancellation.
+
+==== Synchronization in SYCL kernels
+
+Group barriers semantics do not change in the fused kernel and barriers already
+in the unfused kernels are preserved in the fused kernel. Despite this, it is
+worth noting that, in order to introduce synchronization between work items in a
+same work-group executing a fused kernel, a barrier is added between each of the
+kernels being fused. This way, fusing a submission sequence as the one above
+would result in the following one unless the `property::no_barriers` property is
+used:
+
+```c++
+queue.submit([&](handler& cgh){
+  auto accIn1 = bIn1.get_access(cgh);
+  auto accIn2 = bIn2.get_access(cgh);
+  auto accIn3 = bIn3.get_access(cgh);
+  auto accTmp = bTmp.get_access(cgh);
+  auto accOut = bOut.get_access(cgh);
+  cgh.parallel_for<KernelOne>(dataSize,
+  [=](item<1> i) {
+    KernelOne{accIn1, accIn2, accTmp}(i);
+    group_barrier(i.get_group());
+    KernelTwo{accTmp, accIn3, accOut}(i);
+  });
+}
+```
+
+=== Kernel Fusion Limitations
+
+In addition to the cases discussed above, kernel fusion might be canceled by the
+runtime if some undesired scenarios arise. Note that some implementations might
```

Comment:
It seems like there are a good handful of cases that may cause a cancellation. Would it make sense to have a way for the user to query if a "finished" fusion attempt ended with a fusion or a cancellation? I can especially see it be useful for debugging purposes.

Reply:
While there currently is no explicit API to query this, implementations can provide additional information. For example, in the implementation we're currently working on, debugging messages indicating if and why fusion was cancelled will be printed if `SYCL_RT_WARNING_LEVEL >= 1`. 

As the return value from `ext_codeplay_complete_fusion` is occupied by the `event`, it is difficult to return an indication of success from this method.

Reply:
Fair point. I think for now a warning controlled by `SYCL_RT_WARNING_LEVEL >= 1` should be good enough. We may want to document it somewhere so users know, but that's not the job of the proposal. Thanks! 😄